## Reach your audience. Exploring Hacker News Posts
We are going to analise post submissions in popular technology website [Hacker News](https://news.ycombinator.com/) (HN), where people can post articles (posts), vote and comment upon. Our research will focus on the two types of posts, `Ask HN` or `Show HN`.

People post `Ask HN`, to ask rest of the users 

We will tray to answer following questions:
+ Do Ask HN or Show HN receive more comments on average?E91E63
+ Do posts created at a certain time receive more comments on average? 
### Sumary Of The Resluts
* Ask HN receive more comments on average.
* The best time for audience engagement is to post your questions 3 p.m.

Let's see the header and first 5 rows of the data set.

In [1]:
from csv import reader 

# Read in the Hacker News data set
open_file = open('HN_posts_year_to_Sep_26_2016.csv')
read_file = reader(open_file)

#Transfer read file into list of list (table)
hn_list = list(read_file)

#Separate header and body of the table
header = (hn_list[:1]) 
hn = hn_list[1:]

print('header')
print(header)
print('body')
for row in hn[:5]:
    
    print(row)

header
[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']]
body
['12579008', 'You have two days to comment if you want stem cells to be classified as your own', 'http://www.regulations.gov/document?D=FDA-2015-D-3719-0018', '1', '0', 'altstar', '9/26/2016 3:26']
['12579005', 'SQLAR  the SQLite Archiver', 'https://www.sqlite.org/sqlar/doc/trunk/README.md', '1', '0', 'blacksqr', '9/26/2016 3:24']
['12578997', 'What if we just printed a flatscreen television on the side of our boxes?', 'https://medium.com/vanmoof/our-secrets-out-f21c1f03fdc8#.ietxmez43', '1', '0', 'pavel_lishin', '9/26/2016 3:19']
['12578989', 'algorithmic music', 'http://cacm.acm.org/magazines/2011/7/109891-algorithmic-composition/fulltext', '1', '0', 'poindontcare', '9/26/2016 3:16']
['12578979', 'How the Data Vault Enables the Next-Gen Data Warehouse and Data Lake', 'https://www.talend.com/blog/2016/05/12/talend-and-Â\x93the-data-vaultÂ\x94', '1', '0', 'markgainor1', '9/26/2016 3:14']


Lets find  the posts that begin with either "Ask HN" or "Show HN".


In [4]:
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1].lower()
    # print(title)
    if title.startswith("ask hn"):
        ask_posts.append(row)
    elif title.startswith("show hn"):
        show_posts.append(row)
    else:
        other_posts.append(row)
print('Amount of elements in ask_posts list:', len(ask_posts))
print('Amount of elements in show_posts list:', len(show_posts))
print('Amount of elements in other_posts list:', len(other_posts))

Amount of elements in ask_posts list: 9139
Amount of elements in show_posts list: 10158
Amount of elements in other_posts list: 273822


We would like to know total number of comments, and average amount of comments per post in the ask posts:


In [5]:
total_ask_comments = 0
for row in ask_posts:
    comments_no = int(row[4])
    total_ask_comments += comments_no
print('Total Ask comments:', total_ask_comments)
avg_ask_comments = total_ask_comments / len(ask_posts)
print('Average Ask comments:', avg_ask_comments)

Total Ask comments: 94986
Average Ask comments: 10.393478498741656


We would like to know total number of comments, and average amount of comments per post in the show posts:


In [6]:
total_show_comments  = 0
for row in show_posts:
    comments_no = int(row[4])
    total_show_comments += comments_no
print('Total Show comments:', total_show_comments)
avg_show_comments = total_show_comments / len(show_posts)
print('Average Show comments:', avg_show_comments)

Total Show comments: 49633
Average Show comments: 4.886099625910612



Hacker News *ask posts* have average over **10 comments per post**. The *show posts* have almost **5 comments per post**. As we can see ask posts have average over double amount of comments than show posts. It indicates that this type of post tigers much more reactions from the Hacker News readers.

Let's find when people post most of the posts and when readers comment it.

In [7]:
import datetime as dt
result_list = []
for row in ask_posts:
    result_list.append([row[6],int(row[4])])
    
counts_by_hour = {}
comments_by_hour = {}
for row in result_list:
    date_format = "%m/%d/%Y %H:%M"
    date = row[0]
    comments = row[1]
    hour = dt.datetime.strptime(date, date_format).strftime('%H')
    # print(date_dt.hour)
    # hour = date_dt.hour 
    
    if hour in counts_by_hour:
        counts_by_hour[hour] +=1
        comments_by_hour[hour] += comments
    else:
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = comments

We would like to know at what time readers posts most of the comments.

In [8]:
comments_by_hour

{'02': 2996,
 '01': 2089,
 '22': 3372,
 '21': 4500,
 '19': 3954,
 '17': 5547,
 '15': 18525,
 '14': 4972,
 '13': 7245,
 '11': 2797,
 '10': 3013,
 '09': 1477,
 '07': 1585,
 '03': 2154,
 '23': 2297,
 '20': 4462,
 '16': 4466,
 '08': 2362,
 '00': 2277,
 '18': 4877,
 '12': 4234,
 '04': 2360,
 '06': 1587,
 '05': 1838}

We know now that at 3 p.m. (15) people publish the most of the comments (18525), but we what about comments per post? Let's see:  

In [9]:
avg_by_hour  = []
for hour in comments_by_hour:
    posts_no = counts_by_hour[hour]
    comments_no = comments_by_hour[hour]
    avg_by_hour.append([hour, comments_no / posts_no])    
avg_by_hour

[['02', 11.137546468401487],
 ['01', 7.407801418439717],
 ['22', 8.804177545691905],
 ['21', 8.687258687258687],
 ['19', 7.163043478260869],
 ['17', 9.449744463373083],
 ['15', 28.676470588235293],
 ['14', 9.692007797270955],
 ['13', 16.31756756756757],
 ['11', 8.96474358974359],
 ['10', 10.684397163120567],
 ['09', 6.653153153153153],
 ['07', 7.013274336283186],
 ['03', 7.948339483394834],
 ['23', 6.696793002915452],
 ['20', 8.749019607843136],
 ['16', 7.713298791018998],
 ['08', 9.190661478599221],
 ['00', 7.5647840531561465],
 ['18', 7.94299674267101],
 ['12', 12.380116959064328],
 ['04', 9.7119341563786],
 ['06', 6.782051282051282],
 ['05', 8.794258373205741]]

As we can se above 3 p.m is time when is the highest amount of the comments per post it's over 28. In order to better see the result we will swap hor with average comments by hour. 

In [11]:
swap_avg_by_hour = []
for row in avg_by_hour:
    hour = row[0]
    avg_comm = row[1]
    swap_avg_by_hour.append([avg_comm, hour])
swap_avg_by_hour


[[11.137546468401487, '02'],
 [7.407801418439717, '01'],
 [8.804177545691905, '22'],
 [8.687258687258687, '21'],
 [7.163043478260869, '19'],
 [9.449744463373083, '17'],
 [28.676470588235293, '15'],
 [9.692007797270955, '14'],
 [16.31756756756757, '13'],
 [8.96474358974359, '11'],
 [10.684397163120567, '10'],
 [6.653153153153153, '09'],
 [7.013274336283186, '07'],
 [7.948339483394834, '03'],
 [6.696793002915452, '23'],
 [8.749019607843136, '20'],
 [7.713298791018998, '16'],
 [9.190661478599221, '08'],
 [7.5647840531561465, '00'],
 [7.94299674267101, '18'],
 [12.380116959064328, '12'],
 [9.7119341563786, '04'],
 [6.782051282051282, '06'],
 [8.794258373205741, '05']]

And then reverse it, so we can see from highest to lowest amount of comments by hour. As we have seen the most comments per post appears at 3 p.m. and the lowest 9 a.m.

In [12]:
sorted_swap = sorted(swap_avg_by_hour, reverse=True)
sorted_swap


[[28.676470588235293, '15'],
 [16.31756756756757, '13'],
 [12.380116959064328, '12'],
 [11.137546468401487, '02'],
 [10.684397163120567, '10'],
 [9.7119341563786, '04'],
 [9.692007797270955, '14'],
 [9.449744463373083, '17'],
 [9.190661478599221, '08'],
 [8.96474358974359, '11'],
 [8.804177545691905, '22'],
 [8.794258373205741, '05'],
 [8.749019607843136, '20'],
 [8.687258687258687, '21'],
 [7.948339483394834, '03'],
 [7.94299674267101, '18'],
 [7.713298791018998, '16'],
 [7.5647840531561465, '00'],
 [7.407801418439717, '01'],
 [7.163043478260869, '19'],
 [7.013274336283186, '07'],
 [6.782051282051282, '06'],
 [6.696793002915452, '23'],
 [6.653153153153153, '09']]

Top 5 Hours to post Ask HN and to get the highest engagement from the readers.


In [13]:
print("Top 5 Hours for 'Ask HN' Comments")
for average, hour in sorted_swap[:5]:
    hour = dt.datetime.strptime(hour, '%H').strftime('%H:%M')
    print("{}: {:.2f} average comments per post".format(hour, average))

Top 5 Hours for 'Ask HN' Comments
15:00: 28.68 average comments per post
13:00: 16.32 average comments per post
12:00: 12.38 average comments per post
02:00: 11.14 average comments per post
10:00: 10.68 average comments per post


Best time to publish 'Ask HN' is central EU is 9 a.m which corresponds to 3 p.m. CST.

---
TODO:

* Determine if show or ask posts receive more points on average.
* Determine if posts created at a certain time are more likely to receive more points.
* Compare your results to the average number of comments and points other posts receive.
* Use Dataquest's data science project style guide to format your project. https://www.dataquest.io/blog/data-science-project-style-guide/


